File: Desktop/Sources/Chat/DeskPilotApproval.swift
import AppKit
import Foundation

/// The human-approval gate for privileged DeskPilot actions.
///
/// ## Why both halves exist
///
/// An `ask` or `local_confirm` action needs two independent approvals, and
/// Hermes (`runtime_context.wait_for_local_approval`) proceeds only when both
/// arrive and agree on every correlation ID:
///
///   1. **The parent policy socket.** `approval.resolve` is lease-gated: the
///      parent takes the PID of its socket peer, resolves that PID's
///      executable, and checks its code-signing identifier against a closed
///      allowlist. Only this process can pass that. Approving mints a
///      single-use confirmation capability and emits `approval.resolved` to the
///      subscriber — which is Hermes, blocked waiting for exactly that frame.
///
///   2. **The ACP outcome**, which travels back through the bridge.
///
/// This type is the only thing that produces (1), and it produces (2) only
/// after (1) succeeded. Neither alone runs anything: a compromised bridge can
/// forge the ACP outcome all day and Hermes will still deny for want of a
/// parent event, and a parent event with no matching ACP outcome denies too.
///
/// ## What the user is shown
///
/// Everything in the dialog comes from `approval.list`, which the parent
/// answers out of the action registry — the same resolution that authorized
/// the action. The model that asked for this action reads untrusted screen
/// text, web pages, and tool output, so its own account of what it is about to
/// do is not evidence. The bridge deliberately forwards no model text at all
/// (see `deskpilot-approval.ts`); `pendingApprovalID` is a lookup key, and the
/// registry answers the question "what will actually happen".
///
/// The one attacker-influenced surface left is the *values* of the action's
/// inputs. Those are registry-validated but still model-chosen, so they are
/// flattened to a single line each, control characters stripped, length
/// capped, and rendered quoted under an explicit heading — a value cannot lay
/// out a line that reads as another field or as a verdict the policy did not
/// reach.
///
/// ## Fail closed
///
/// Denial is the outcome for: no live UI lease; a `pendingApprovalID` the
/// parent is not holding; an item that will not fully parse; an item already
/// past its expiry; an expiry that lapses while the dialog is open; the item
/// changing between being shown and being approved; a refused or failed
/// `approval.resolve`; and a dismissed dialog. The user not answering is not
/// consent — the parent expires the pending approval on its own and the bridge
/// answers the stranded ACP request.
@MainActor
final class DeskPilotApprovalCoordinator {

  static let shared = DeskPilotApprovalCoordinator()

  /// One permission request as the bridge described it: correlation and expiry.
  /// Note what is absent — there is no action name here, because the bridge is
  /// not a trustworthy source for one.
  struct Request: Equatable, Sendable {
    let routeID: String
    let sessionID: String
    let permissionRequestID: String
    let pendingApprovalID: String
    let expiresAt: Date
  }

  /// Sends the ACP half back down the bridge. Set by `ACPBridge`.
  var respond: (@Sendable (Request, Bool) -> Void)?

  /// Requests waiting behind the one on screen. Serialised deliberately: two
  /// consent dialogs at once is how a user approves the wrong one.
  private var queued: [Request] = []
  private var presenting = false
  private var closed = Set<String>()

  private init() {}

  // MARK: - Inbound

  func handle(_ request: Request) {
    guard !closed.contains(request.permissionRequestID) else {
      // The bridge already told us this request died; nothing to consent to.
      closed.remove(request.permissionRequestID)
      deny(request, why: "request already closed")
      return
    }
    queued.append(request)
    drain()
  }

  /// The bridge says this request is gone (expired, cancelled, transport lost).
  /// Drop it so it can never be presented or approved after the fact.
  func close(permissionRequestID: String) {
    queued.removeAll { $0.permissionRequestID == permissionRequestID }
    closed.insert(permissionRequestID)
  }

  // MARK: - Queue

  private func drain() {
    guard !presenting, !queued.isEmpty else { return }
    let request = queued.removeFirst()
    presenting = true
    lookUpThenPresent(request)
  }

  private func finish(_ request: Request, approved: Bool) {
    respond?(request, approved)
    presenting = false
    drain()
  }

  private func deny(_ request: Request, why: String) {
    fputs("[deskpilot] approval denied for \(request.pendingApprovalID): \(why)\n", stderr)
    finish(request, approved: false)
  }

  // MARK: - Look-up

  /// Ask the parent what this pending approval actually is, off the main
  /// thread, then present. A request the parent is not holding is denied
  /// without ever being shown: there would be nothing truthful to show.
  private func lookUpThenPresent(_ request: Request) {
    let pendingID = request.pendingApprovalID
    Task {
      let found = await Task.detached(priority: .userInitiated) {
        Self.lookUp(pendingApprovalID: pendingID)
      }.value

      guard let action = found else {
        deny(request, why: "parent is not holding this pending approval")
        return
      }
      guard action.expiresAt > Date() else {
        deny(request, why: "already expired at look-up")
        return
      }
      guard !closed.contains(request.permissionRequestID) else {
        closed.remove(request.permissionRequestID)
        deny(request, why: "closed while being looked up")
        return
      }
      present(request, action)
    }
  }

  nonisolated private static func lookUp(pendingApprovalID: String) -> DeskPilotPendingAction? {
    guard let lease = DeskPilotUILease.lease() else {
      fputs("[deskpilot] approval: no live UI lease\n", stderr)
      return nil
    }
    do {
      return try DeskPilotApprovalService.find(lease: lease, pendingApprovalID: pendingApprovalID)
    } catch {
      fputs("[deskpilot] approval: look-up failed: \(error)\n", stderr)
      return nil
    }
  }

  // MARK: - Presentation

  private func present(_ request: Request, _ action: DeskPilotPendingAction) {
    let alert = NSAlert()
    alert.alertStyle = .critical
    alert.messageText = "Allow “\(action.actionID)” to run?"
    alert.informativeText = Self.body(for: action, expiresAt: request.expiresAt)

    // Deny is added first, so it is the default and Return cannot approve.
    alert.addButton(withTitle: "Deny")
    alert.addButton(withTitle: "Allow Once")

    NSApp.activate(ignoringOtherApps: true)
    alert.window.level = .modalPanel
    let response = alert.runModal()
    let approved = response == .alertSecondButtonReturn

    guard approved else {
      recordDenial(request)
      return
    }
    confirm(request, action)
  }

  /// The dialog body, assembled entirely from registry and policy fields.
  static func body(for action: DeskPilotPendingAction, expiresAt: Date) -> String {
    var lines: [String] = []
    lines.append("DeskPilot is asking to run a privileged action on this Mac.")
    lines.append("")
    lines.append("Action:   \(action.actionID) (version \(action.actionVersion))")
    lines.append("Risk:     \(action.risk)")
    lines.append("Policy:   \(action.verdict) — \(action.ruleID)")
    lines.append("Reason:   \(action.reason)")
    lines.append("Source:   \(action.entryPoint)\(action.sender.map { " (\($0))" } ?? "")")
    lines.append("")
    if action.displayInputs.isEmpty {
      lines.append("Inputs:   (none)")
    } else {
      lines.append("Inputs, as validated by the action registry:")
      for (key, value) in action.displayInputs {
        lines.append("  • \(key) = \u{201C}\(value)\u{201D}")
      }
    }
    lines.append("")
    lines.append("Trace:    \(action.traceID)")
    lines.append("Digest:   \(action.actionDigest)")
    lines.append("Expires:  \(DeskPilotUILease.rfc3339(expiresAt))")
    lines.append("")
    lines.append(
      "Allowing permits this action once, with exactly these inputs. "
        + "It does not grant anything else, and it cannot be reused.")
    return lines.joined(separator: "\n")
  }

  // MARK: - Resolution

  /// Approve: parent first, then ACP.
  ///
  /// The order is forced by Hermes, which blocks on `approval.subscribe` for
  /// the parent event before it ever reads the ACP outcome. It is also the safe
  /// order: if the parent refuses, nothing has been told "allowed" yet.
  ///
  /// Before resolving, the item is read back and compared field-for-field with
  /// what was on screen. That closes the window between showing and clicking —
  /// if the pending approval is no longer exactly the action whose inputs the
  /// user read, their consent does not apply to it.
  private func confirm(_ request: Request, _ action: DeskPilotPendingAction) {
    Task {
      let outcome = await Task.detached(priority: .userInitiated) {
        Self.resolveApproval(request: request, action: action)
      }.value

      if outcome.approved {
        fputs("[deskpilot] approved \(action.actionID) once (pending \(request.pendingApprovalID))\n", stderr)
        finish(request, approved: true)
      } else {
        deny(request, why: outcome.why)
      }
    }
  }

  nonisolated private static func resolveApproval(
    request: Request, action: DeskPilotPendingAction
  ) -> (approved: Bool, why: String) {
    guard let lease = DeskPilotUILease.lease() else {
      return (false, "no live UI lease at confirmation")
    }
    do {
      guard let fresh = try DeskPilotApprovalService.find(
        lease: lease, pendingApprovalID: request.pendingApprovalID)
      else { return (false, "pending approval vanished before it could be resolved") }

      guard fresh == action else {
        return (false, "pending approval changed between being shown and being approved")
      }
      guard fresh.expiresAt > Date() else {
        return (false, "expired while the dialog was open")
      }
      guard try DeskPilotApprovalService.resolve(
        lease: lease, pendingApprovalID: request.pendingApprovalID, approve: true)
      else { return (false, "parent did not accept the resolution") }

      return (true, "approved")
    } catch {
      return (false, "parent refused or was unreachable: \(error)")
    }
  }

  /// Deny: tell the parent so it resolves and revokes, then tell the bridge.
  /// The ACP denial is sent whether or not the parent call succeeded — a
  /// failure to record the denial must never read as an approval.
  private func recordDenial(_ request: Request) {
    let pendingID = request.pendingApprovalID
    Task {
      await Task.detached(priority: .userInitiated) {
        guard let lease = DeskPilotUILease.lease() else { return }
        _ = try? DeskPilotApprovalService.resolve(
          lease: lease, pendingApprovalID: pendingID, approve: false)
      }.value
      deny(request, why: "denied by the user")
    }
  }
}

File: acp-bridge/src/deskpilot-approval.ts
/**
 * The bridge half of DeskPilot's human-approval path.
 *
 * An `ask` or `local_confirm` action needs two independent approvals before it
 * runs, and Hermes (`runtime_context.wait_for_local_approval`) refuses to
 * proceed without both:
 *
 *   1. the **ACP** `session/request_permission` outcome, which travels back
 *      down this bridge, and
 *   2. the **parent policy socket**'s `approval.resolved` event, which the app
 *      triggers directly by calling `approval.resolve` with its verified UI
 *      lease — a socket this process cannot satisfy, because the parent takes
 *      the peer PID and checks that PID's code signature.
 *
 * That split is the point. This module carries leg 1 only. It cannot mint leg 2
 * and must never be able to: if the bridge were subverted it could reply
 * "allowed" here all day and Hermes would still deny, because no matching
 * parent event exists. Equally, a parent event alone is not enough — Hermes
 * waits on this ACP outcome too.
 *
 * ## What the app is told, and what it is deliberately not told
 *
 * `describe()` returns correlation IDs and an expiry, and nothing else. The
 * inbound `session/request_permission` frame also carries a `title` and a
 * `rawInput` built by Hermes from a model turn, and that model reads untrusted
 * screen text, web pages, and tool output. Forwarding any of it would let the
 * model write the words a human reads before consenting. So the app gets a
 * `pendingApprovalID` and uses it as a lookup key against the parent policy
 * server's `approval.list`, which answers from the action registry: the
 * resolved action ID, its version, its validated inputs, and the policy
 * verdict. The human is shown the registry's account of what will happen, not
 * the model's.
 *
 * ## Fail-closed inventory
 *
 * Every one of these denies, and none of them is recoverable by retrying:
 * an uncorrelated or malformed `_meta.deskpilot`; a response naming an unknown
 * permission request; a response whose routeID, sessionID, or pendingApprovalID
 * disagrees with the record; a response arriving at or after `expiresAt`; a
 * decision that is not exactly `allow_once` or `deny`; a replayed response; a
 * cancelled session; and transport loss (see `drain`). The absence of a
 * response is a denial too — `GenericACPProvider` runs its own expiry timer and
 * answers `cancelled` when it fires.
 */

import type { PermissionHandle } from "./stdio-provider.js";

/** The exact `_meta.deskpilot` block Hermes attaches to a permission request. */
export interface ApprovalCorrelation {
  routeID: string;
  sessionID: string;
  permissionRequestID: string;
  pendingApprovalID: string;
  expiresAt: string;
}

export type DenyReason =
  | "malformed"
  | "unknown_request"
  | "correlation_mismatch"
  | "expired"
  | "denied";

export type ApprovalVerdict =
  | { allowed: true; handle: PermissionHandle }
  | { allowed: false; reason: DenyReason; handle?: PermissionHandle };

interface Record_ {
  correlation: ApprovalCorrelation;
  handle: PermissionHandle;
  expiresAtMs: number;
}

const CORRELATION_KEYS = [
  "routeID",
  "sessionID",
  "permissionRequestID",
  "pendingApprovalID",
  "expiresAt",
] as const;

const RESPONSE_KEYS = [
  "type",
  "routeID",
  "sessionID",
  "permissionRequestID",
  "pendingApprovalID",
  "decision",
] as const;

/** Lowercase canonical UUID. `routeID` and `pendingApprovalID` are minted by
 *  Hermes and the parent respectively; nothing else is a legitimate value. */
const UUID = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/;

function nonEmptyString(value: unknown): value is string {
  return typeof value === "string" && value.length > 0;
}

function exactKeys(value: unknown, keys: readonly string[]): value is Record<string, unknown> {
  if (typeof value !== "object" || value === null || Array.isArray(value)) return false;
  const actual = Object.keys(value as Record<string, unknown>);
  return actual.length === keys.length && keys.every((key) => actual.includes(key));
}

/**
 * Extract the correlation from an inbound `session/request_permission` frame,
 * or null if the frame does not carry an exact, well-typed one. Null means
 * deny: an uncorrelated request can never be matched to a parent approval, so
 * approving it would be approving something unidentifiable.
 */
export function readCorrelation(frame: unknown): ApprovalCorrelation | null {
  const params = (frame as { params?: unknown } | undefined)?.params;
  if (typeof params !== "object" || params === null) return null;
  const meta = (params as { _meta?: unknown })._meta;
  if (typeof meta !== "object" || meta === null) return null;
  const deskpilot = (meta as { deskpilot?: unknown }).deskpilot;
  if (!exactKeys(deskpilot, CORRELATION_KEYS)) return null;

  for (const key of CORRELATION_KEYS) {
    if (!nonEmptyString(deskpilot[key])) return null;
  }
  if (!UUID.test(deskpilot.routeID as string)) return null;
  if (!UUID.test(deskpilot.pendingApprovalID as string)) return null;
  if (!Number.isFinite(Date.parse(deskpilot.expiresAt as string))) return null;

  return {
    routeID: deskpilot.routeID as string,
    sessionID: deskpilot.sessionID as string,
    permissionRequestID: deskpilot.permissionRequestID as string,
    pendingApprovalID: deskpilot.pendingApprovalID as string,
    expiresAt: deskpilot.expiresAt as string,
  };
}

/**
 * The bridge's pending-approval table. One entry per in-flight
 * `session/request_permission`, keyed by the permission request ID Hermes
 * generated. Entries leave the table exactly once — resolved, expired,
 * cancelled, or drained — so no approval can be replayed.
 */
export class DeskPilotApprovalStore {
  private readonly records = new Map<string, Record_>();

  get size(): number {
    return this.records.size;
  }

  /** Register a request. Refuses a duplicate ID rather than overwriting: an
   *  overwrite would strand the first ACP request without a reply, and would
   *  let a second request inherit the first's pending human decision. */
  open(correlation: ApprovalCorrelation, handle: PermissionHandle): boolean {
    const key = correlation.permissionRequestID;
    if (this.records.has(key)) return false;
    if (handle.permissionRequestID !== key || handle.sessionId !== correlation.sessionID) return false;
    this.records.set(key, { correlation, handle, expiresAtMs: Date.parse(correlation.expiresAt) });
    return true;
  }

  /** What the app is allowed to see: correlation and expiry, no model text. */
  describe(permissionRequestID: string): Omit<ApprovalCorrelation, never> | Record<string, never> {
    const record = this.records.get(permissionRequestID);
    if (!record) return {} as Record<string, never>;
    const { routeID, sessionID, pendingApprovalID, expiresAt } = record.correlation;
    return { routeID, sessionID, permissionRequestID, pendingApprovalID, expiresAt };
  }

  /**
   * Apply the app's decision. Consumes the record on every outcome except a
   * malformed frame — a garbled or hostile message must not be able to burn a
   * legitimate pending approval and strand the turn.
   */
  resolve(response: unknown, nowMs: number): ApprovalVerdict {
    if (!exactKeys(response, RESPONSE_KEYS)) return { allowed: false, reason: "malformed" };
    for (const key of RESPONSE_KEYS) {
      if (!nonEmptyString(response[key])) return { allowed: false, reason: "malformed" };
    }
    // The whole wire frame is validated, `type` included, so an extra or
    // renamed field is a rejection rather than something quietly ignored.
    if (response.type !== "deskpilot_permission_response") {
      return { allowed: false, reason: "malformed" };
    }

    const key = response.permissionRequestID as string;
    const record = this.records.get(key);
    if (!record) return { allowed: false, reason: "unknown_request" };

    const { correlation, handle } = record;
    if (
      response.routeID !== correlation.routeID ||
      response.sessionID !== correlation.sessionID ||
      response.pendingApprovalID !== correlation.pendingApprovalID
    ) {
      // The record survives: a mismatched reply is somebody else's, and the
      // real one may still arrive before expiry.
      return { allowed: false, reason: "correlation_mismatch" };
    }

    this.records.delete(key);

    // Expiry is checked after correlation and before the decision, so a stale
    // click cannot be rescued by carrying the right IDs.
    if (nowMs >= record.expiresAtMs) return { allowed: false, reason: "expired", handle };
    if (response.decision !== "allow_once") return { allowed: false, reason: "denied", handle };
    return { allowed: true, handle };
  }

  /** Forget one request without deciding it (its ACP request was answered by
   *  the provider's own expiry timer). */
  forget(permissionRequestID: string): void {
    this.records.delete(permissionRequestID);
  }

  /** Drop every approval belonging to a cancelled session. */
  cancelSession(sessionID: string): Record_[] {
    const dropped: Record_[] = [];
    for (const [key, record] of this.records) {
      if (record.correlation.sessionID !== sessionID) continue;
      this.records.delete(key);
      dropped.push(record);
    }
    return dropped;
  }

  /** Empty the table, handing back what was in it so each can be denied. */
  drain(): Record_[] {
    const all = [...this.records.values()];
    this.records.clear();
    return all;
  }
}

/**
 * The ACP outcome for an approved request.
 *
 * The discriminator is `"selected"`, not `"allowed"` — `acp.schema.AllowedOutcome`
 * declares `outcome: Literal["selected"]`. Sending `"allowed"` produced a
 * response Hermes could not read as an approval, so it denied with "approval
 * capability required" even though the parent had already resolved. That was
 * the dual-authority gate behaving correctly on a bridge that was wrong, which
 * is exactly why the ACP leg is not permitted to be a formality.
 *
 * `allow_once` is the only option DeskPilot offers: there is no allow-always
 * and no session-scoped grant, so no approval can cover an action the user was
 * not shown.
 */
export const ALLOW_ONCE_OUTCOME = { outcome: "selected", optionId: "allow_once" } as const;

/** The ACP outcome for everything else. `acp.schema.DeniedOutcome` is
 *  `outcome: Literal["cancelled"]`. */
export const DENY_OUTCOME = { outcome: "cancelled" } as const;
